[Bug]: terraform-provider-vault_v4.6.0_x5 crashes on vault_pki_secret_backend_role apply

### Terraform Core Version

1.6.5

### Terraform Vault Provider Version

4.6.0

### Vault Server Version

1.10.0

### Affected Resource(s)

`vault_pki_secret_backend_role`

### Expected Behavior

No crash.

### Actual Behavior

Plugin version 4.6.0 crashes, version 4.5.0 works fine.

### Relevant Error/Panic Output Snippet

```shell
╷
│ Error: Request cancelled
│
│   with vault_pki_secret_backend_role.repro,
│   on repro.tf line 21, in resource "vault_pki_secret_backend_role" "repro":
│   21: resource "vault_pki_secret_backend_role" "repro" {
│
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.
╵

Stack trace from the terraform-provider-vault_v4.6.0_x5 plugin:

panic: interface conversion: interface {} is nil, not []interface {}

goroutine 85 [running]:
github.com/hashicorp/terraform-provider-vault/vault.pkiSecretBackendRoleRead({0x14000a3ea00?, 0x1400074e690?}, 0x14000a3ea00, {0x103c91400, 0x14000313c40})
        github.com/hashicorp/terraform-provider-vault/vault/resource_pki_secret_backend_role.go:578 +0xdc0
github.com/hashicorp/terraform-provider-vault/vault.pkiSecretBackendRoleCreate({0x103d39570, 0x1400068e460}, 0x14000a3ea00, {0x103c91400, 0x14000313c40})
        github.com/hashicorp/terraform-provider-vault/vault/resource_pki_secret_backend_role.go:536 +0x948
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x140006f1260, {0x103d394c8, 0x14000b926f0}, 0x14000a3ea00, {0x103c91400, 0x14000313c40})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:778 +0xe4
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x140006f1260, {0x103d394c8, 0x14000b926f0}, 0x14000a6edd0, 0x14000a3e780, {0x103c91400, 0x14000313c40})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:909 +0x874
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x1400087ed08, {0x103d394c8?, 0x14000b92630?}, 0x14000a5ddb0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/grpc_provider.go:1074 +0xaa4
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x140004fc1e0, {0x103d394c8?, 0x14000a6bdd0?}, 0x1400066be30)
        github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/tf5server/server.go:859 +0x3c0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x103cdca40, 0x140004fc1e0}, {0x103d394c8, 0x14000a6bdd0}, 0x14000a3e480, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:503 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x140005e6400, {0x103d394c8, 0x14000a6bd40}, {0x103d40a40, 0x14000838680}, 0x14000b8cc60, 0x14000725650, 0x10474be98, 0x0)
        google.golang.org/grpc@v1.61.1/server.go:1385 +0xb40
google.golang.org/grpc.(*Server).handleStream(0x140005e6400, {0x103d40a40, 0x14000838680}, 0x14000b8cc60)
        google.golang.org/grpc@v1.61.1/server.go:1796 +0xc00
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.61.1/server.go:1029 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 24
        google.golang.org/grpc@v1.61.1/server.go:1040 +0x13c

Error: The terraform-provider-vault_v4.6.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Terraform Configuration Files

```tf
terraform {
  required_providers {
    vault = {
      source  = "hashicorp/vault"
      version = "~> 4.6.0"
    }
  }

  backend "local" {
    path = "terraform.tfstate"
  }
}

resource "vault_mount" "repro" {
  path                      = "pki/repro-plugin-crash"
  type                      = "pki"
  default_lease_ttl_seconds = 3600
  max_lease_ttl_seconds     = 86400
}

resource "vault_pki_secret_backend_role" "repro" {
  backend = vault_mount.repro.path
  name = "repro"
  ttl              = 3600
  allow_ip_sans    = true
  key_type         = "rsa"
  key_bits         = 4096
  allowed_domains  = ["example.com", "my.domain"]
  allow_subdomains = true
}
```

### Steps to Reproduce

Use the provided repro terraform file, run:
```sh
terraform apply
```

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

Looks like the crash happens at [resource_pki_secret_backend_role.go:578](https://github.com/hashicorp/terraform-provider-vault/blob/4b26c7a2c39ea4012282efa0570533c052130b2a/vault/resource_pki_secret_backend_role.go#L578).

### References

_No response_

### Would you like to implement a fix?

No

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,10 @@
 ## Unreleased
 
+BUGS:
+
+* Do not panic on Vault PKI roles without the cn_validations field: ([#2398](https://github.com/hashicorp/terraform-provider-vault/pull/2398))
+
+
 ## 4.6.0 (Jan 15, 2025)
 
 FEATURES:

--- a/util/util.go
+++ b/util/util.go
@@ -14,6 +14,7 @@ import (
 	"time"
 
 	"github.com/cenkalti/backoff/v4"
+	"github.com/hashicorp/go-secure-stdlib/parseutil"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/vault/api"
 )
@@ -449,4 +450,41 @@ func RetryWrite(client *api.Client, path string, data map[string]interface{}, re
 		func(err error, duration time.Duration) {
 			log.Printf("[WARN] Writing to path %q failed, retrying in %s", path, duration)
 		})
+}
+
+// GetStringSliceFromSecret will return a string slice from the secret data within the provided field name if it exists.
+// The bool return value will be false if the field does not exist or is not a string slice. It will be true if the field
+// exists and was an empty slice.
+func GetStringSliceFromSecret(secret *api.Secret, fieldName string) ([]string, bool) {
+	if secret == nil || secret.Data == nil {
+		return nil, false
+	}
+
+	rawVal, exists := secret.Data[fieldName]
+	if !exists {
+		return nil, false
+	}
+
+	rv := reflect.ValueOf(rawVal)
+	switch rv.Kind() {
+	case reflect.Slice:
+		if rv.IsNil() {
+			return nil, false
+		}
+	case reflect.Array:
+	default:
+		return nil, false
+	}
+
+	output := make([]string, rv.Len())
+
+	for i := 0; i < rv.Len(); i++ {
+		myStr, err := parseutil.ParseString(rv.Index(i).Interface())
+		if err != nil {
+			return nil, false
+		}
+		output[i] = myStr
+	}
+
+	return output, true
 }

--- a/util/util_test.go
+++ b/util/util_test.go
@@ -871,3 +871,39 @@ func TestRetryWrite(t *testing.T) {
 		})
 	}
 }
+
+func TestGetStringSliceFromSecret(t *testing.T) {
+	fieldName := "foo"
+	var testArray [2]string
+	testArray[0] = "1"
+	testArray[1] = "2"
+
+	tests := []struct {
+		name       string
+		secretData map[string]interface{}
+		want       []string
+		wantOk     bool
+	}{
+		{"nil-data", nil, nil, false},
+		{"field-missing", map[string]interface{}{}, nil, false},
+		{"nil-element", map[string]interface{}{fieldName: nil}, nil, false},
+		{"not-a-slice", map[string]interface{}{fieldName: "test-value"}, nil, false},
+		{"array-val", map[string]interface{}{fieldName: testArray}, []string{"1", "2"}, true},
+		{"mixed-slice", map[string]interface{}{fieldName: []interface{}{1, "2"}}, []string{"1", "2"}, true},
+		{"int-slice", map[string]interface{}{fieldName: []int{1, 2}}, []string{"1", "2"}, true},
+		{"string-slice", map[string]interface{}{fieldName: []string{"test1", "test2"}}, []string{"test1", "test2"}, true},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			secret := &api.Secret{Data: tt.secretData}
+			got, gotOk := GetStringSliceFromSecret(secret, fieldName)
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("GetStringSliceFromSecret() got = %v, want %v", got, tt.want)
+			}
+			if gotOk != tt.wantOk {
+				t.Errorf("GetStringSliceFromSecret() gotOk = %v, wantOk %v", gotOk, tt.wantOk)
+			}
+		})
+	}
+}

--- a/vault/resource_pki_secret_backend_role.go
+++ b/vault/resource_pki_secret_backend_role.go
@@ -14,10 +14,10 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
-
 	"github.com/hashicorp/terraform-provider-vault/internal/consts"
 	"github.com/hashicorp/terraform-provider-vault/internal/pki"
 	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+	"github.com/hashicorp/terraform-provider-vault/util"
 )
 
 var (
@@ -575,10 +575,10 @@ func pkiSecretBackendRoleRead(_ context.Context, d *schema.ResourceData, meta in
 	listFields := append(pkiSecretListFields, consts.FieldKeyUsage)
 	// handle TypeList
 	for _, k := range listFields {
-		list := expandStringSlice(secret.Data[k].([]interface{}))
-
-		if len(list) > 0 {
-			d.Set(k, list)
+		if list, ok := util.GetStringSliceFromSecret(secret, k); ok {
+			if len(list) > 0 {
+				d.Set(k, list)
+			}
 		}
 	}
 

